packages/vscode-integration/src/types.ts

```diff
@@ -0,0 +1,196 @@
+
+import * as vscode from 'vscode';
+
+/**
+ * Any clients registered on the GLSP VSCode integration need to implement this
+ * interface.
+ */
+export interface GlspVscodeClient<D extends vscode.CustomDocument = vscode.CustomDocument> {
+
+    /**
+     * A unique identifier for the client/panel with which the client will be registered
+     * on the server.
+     */
+    readonly clientId: string;
+
+    /**
+     * The webview belonging to the client.
+     */
+    readonly webviewPanel: vscode.WebviewPanel;
+
+    /**
+     * The document object belonging to the client;
+     */
+    readonly document: D;
+
+    /**
+     * This event emitter is used by the VSCode integration to pass messages/actions
+     * to the client. These messages can come from the server or the VSCode integration
+     * itself.
+     *
+     * You should subscribe to the attached event and pass contents of the event
+     * to the webview.
+     *
+     * Use the properties `onBeforeReceiveMessageFromServer` and `onBeforePropagateMessageToClient`
+     * of the GlspVscodeAdapter in order to control what messages are propagated
+     * and processed.
+     */
+    readonly onClientReceiveEmitter: vscode.EventEmitter<unknown>;
+
+    /**
+     * This event is used to notify the VSCode integration about messages from the
+     * client.
+     *
+     * Fire this event with the message the client wants to send to the server.
+     *
+     * Use the properties `onBeforeReceiveMessageFromClient` and `onBeforePropagateMessageToServer`
+     * of the GlspVscodeAdapter in order to control what messages are propagated
+     * and processed.
+     */
+    readonly onClientSend: vscode.Event<unknown>;
+}
+
+/**
+ * The server or server wrapper used by the VSCode integration needs to implement
+ * this interface.
+ */
+export interface GlspVscodeServer {
+
+    /**
+     * An event emitter used by the VSCode extension to send messages to the server.
+     * You should listen to the event attached to this emitter to receive messages
+     * from the client/VSCode integration and send it to the server.
+     *
+     * Use the properties `onBeforeReceiveMessageFromClient` and `onBeforePropagateMessageToServer`
+     * of the GlspVscodeAdapter in order to control what messages are propagated
+     * and processed.
+     */
+    readonly onServerReceiveEmitter: vscode.EventEmitter<unknown>;
+
+    /**
+     * An event the VSCode integration uses to receive messages from the server.
+     * The messages are then propagated to the client or processed by the VSCode
+     * integration to provide functionality.
+     *
+     * Fire this event with the message you want to send to the client.
+     *
+     * Use the properties `onBeforeReceiveMessageFromServer` and `onBeforePropagateMessageToClient`
+     * of the GlspVscodeAdapter in order to control what messages are propagated
+     * and processed.
+     */
+    readonly onServerSend: vscode.Event<unknown>;
```

Comment:
I understand the confusion. I spent way too much time coming up with names here because the events and event emitters do such similar things. In this case I wouldn't name it `*Recieved` because this event should be fired when the server(-process) sent a message and not when it receives one.

I settled with `onServerMessage`: 26e430709fab8d764971bf477ba694e2595aa6a2

Feel free to suggest further names if you feel like this one doesn't fit either.